test/vm/unit/uint256_tests.cpp

```diff
@@ -436,15 +453,64 @@ TEST(uint256, arithmetic)
                 }
                 ASSERT_EQ(
                     addmod(x, y, z),
-                    uint256_t(
-                        ::intx::addmod(x.to_intx(), y.to_intx(), z.to_intx())));
+                    from_intx(
+                        ::intx::addmod(to_intx(x), to_intx(y), to_intx(z))));
                 ASSERT_EQ(
                     mulmod(x, y, z),
-                    uint256_t(
-                        ::intx::mulmod(x.to_intx(), y.to_intx(), z.to_intx())));
+                    from_intx(
+                        ::intx::mulmod(to_intx(x), to_intx(y), to_intx(z))));
             }
         }
-        ASSERT_EQ(-x, uint256_t(-(x.to_intx())));
+        ASSERT_EQ(-x, from_intx(-(to_intx(x))));
+    }
+}
+
+template <size_t R, size_t M, size_t N>
+void check_truncating_mul(
+    words_t<M> const &x, words_t<N> const &y,
+    words_t<M + N> const &full) noexcept
+{
+    auto const prod_rt = truncating_mul_runtime<R>(x, y);
+    auto const prod_ce = truncating_mul_constexpr<R>(x, y);
+    for (size_t i = 0; i < R; i++) {
+        ASSERT_EQ(prod_rt[i], full[i]);
+        ASSERT_EQ(prod_ce[i], full[i]);
+    }
+}
+
+TEST(uint256, multiplication)
+{
+    for (auto const &x : test_inputs) {
+        for (auto const &y : test_inputs) {
+            auto const intx_product =
+                std::bit_cast<words_t<2 * uint256_t::num_words>>(
+                    ::intx::umul(to_intx(x), to_intx(y)));
+
+            auto const full_product_rt =
+                truncating_mul_runtime<2 * uint256_t::num_words>(
+                    x.as_words(), y.as_words());
+            ASSERT_EQ(full_product_rt, intx_product);
+
+            auto const full_product_ce =
+                truncating_mul_constexpr<2 * uint256_t::num_words>(
+                    x.as_words(), y.as_words());
+            ASSERT_EQ(full_product_ce, intx_product);
+
+            check_truncating_mul<1>(
```

Comment:
Perhaps it would be good to have unit tests for the cases `(M = 1,N = 1, R = 1..2)`, `(M = 2, N = 1, R = 1..3)`, `(M = 1, N = 2, R = 1..3)`.